functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,170 @@
+	for key, val := range t.ConfigMap {
+		switch key {
+		case "name":
+			t.Image.Name = val
+		case "newName":
+			t.Image.NewName = val
+		case "newTag":
+			t.Image.NewTag = val
+		case "digest":
+			t.Image.Digest = val
+		default:
+			return fmt.Errorf("ConfigMap has wrong field name %v", key)
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	if t.Image.NewTag != "" && t.Image.Digest != "" {
+		return fmt.Errorf("image `newTag` and `digest` both set, set only one")
+	}
+	return nil
+}
+
+func (t *SetImage) setPodSpecContainers(o *fn.KubeObject) error {
+	podSpec := o.GetMap("spec").GetMap("template").GetMap("spec")
+	for _, vecObj := range podSpec.GetSlice("containers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	for _, vecObj := range podSpec.GetSlice("iniContainers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *SetImage) setPodContainers(o *fn.KubeObject) error {
+	spec := o.GetMap("spec")
+	for _, vecObj := range spec.GetSlice("containers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	for _, vecObj := range spec.GetSlice("iniContainers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *SetImage) hasPodSpecContainers(o *fn.KubeObject) bool {
+	spec := o.GetMap("spec")
+	if spec == nil {
+		return false
+	}
+	template := spec.GetMap("template")
+	if template == nil {
+		return false
+	}
+	podSpec := template.GetMap("spec")
+	return podSpec != nil
+}
+
+func (t *SetImage) hasPodContainers(o *fn.KubeObject) bool {
+	spec := o.GetMap("spec")
+	return spec != nil
+}
+
+// getNewImageName return the new name for image field
+func getNewImageName(oldValue string, newImage *image_util.Image) string {
+	name, tag := image_util.Split(oldValue)
+	if newImage.NewName != "" {
+		name = newImage.NewName
+	}
+	if newImage.NewTag != "" {
+		tag = ":" + newImage.NewTag
+	}
+	if newImage.Digest != "" {
+		tag = "@" + newImage.Digest
+	}
+	newName := name + tag
+	return newName
+}
+
+// updateImages update the image for a given fieldpath
+func (t *SetImage) updateImages(o *fn.SubObject, newImage *image_util.Image, parentO *fn.KubeObject) error {
+	oldValue := o.NestedStringOrDie("image")
+	if !image_util.IsImageMatched(oldValue, newImage.Name) {
+		return nil
+	}
+	newName := getNewImageName(oldValue, newImage)
+	err := o.SetNestedString(newName, "image")
+
+	msg := fmt.Sprintf("updated image from %v to %v", oldValue, newName)
```

Comment:
is that possible the old and new image names are the same?  If so, we should not update